src/samVisualize/graphGeneration/cfnTemplateGraphGenerator.ts

```diff
@@ -0,0 +1,150 @@
+ * @param graph The Graph in which links are created. It is modified, due to traverse being recursive. But this Graph is not visible from outside the module, so there are no external side effects
+ * @param currentObj The object wich is currently being searched for link keys
+ * @param parentNodeName The name of the node representing the current parent node, which any found links will be added to.
+ */
+function traverse(graph: Graph, currentObj: Record<string, any>, parentNodeName: string): void {
+    // Cannot traverse on a string or number literal, no links will be found.
+    if (!_.isObjectLike(currentObj)) {
+        return
+    }
+    for (const [key, value] of Object.entries(currentObj)) {
+        switch (key) {
+            //  "DepenedsOn" can point to a single string or a list of strings. Here we wish to capture all resources in a list, or just a single resource.
+            //  https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-dependson.html
+            case LinkTypes.DependsOn:
+                if (Array.isArray(value)) {
+                    for (const destNodeName of value) {
+                        graph.createLink(parentNodeName, destNodeName, LinkTypes.DependsOn)
+                    }
+                } else {
+                    graph.createLink(parentNodeName, value, LinkTypes.DependsOn)
+                }
+                break
+
+            //  Once the YAML is parsed, a "GetAtt" key will point to an array with two elements.
+            //  We only want the logicalNameOfResource, which lies in the first element of the tuple.
+            //  https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-getatt.html
+            case LinkTypes.GetAtt:
+                graph.createLink(parentNodeName, value[0], LinkTypes.GetAtt)
+                break
+
+            //  Adding a single link, from a "Ref" key
+            //  https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-ref.html
+            case LinkTypes.Ref:
+                graph.createLink(parentNodeName, value, LinkTypes.Ref)
+                break
+
+            //  Extracting an link out of a substitution
+            //  A sub link can point to an array, or a single string.
+            //  We can immediately extract the substituion if it points to a string, otherwise we must continue traversing
+            //  https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-sub.html
+            case LinkTypes.Sub:
+                if (_.isString(value)) {
+                    const substitutions = extractSubstitution(value)
+                    for (const destNodeName of substitutions) {
+                        graph.createLink(parentNodeName, destNodeName, LinkTypes.Sub)
+                    }
+                } else {
+                    // Key must point to an array
+                    for (const element of value) {
```

Comment:
For the 'array' case, isn't the first element a string with substitutions? In that case, the first element would be ignored in the traverse call, even though it may have valid substitutions. A test case for this would be nice.

Reply:
The case of `Fn::Sub` pointing to an array with a first element string happens when you want to substitute variables into that string. I believe it makes sense to ignore the string itself, because the variables being substituted get defined in the second object element, the `resource map`. This resource map is what we care about in this case because that's where the links to other resources will be made. There is an example of this in one of the tests, 
```x```